### ISSUE TITLE
Stabilize the Table of Contents Block

### ISSUE THREAD

**Issue by priethor** [WordPress/gutenberg#42229]
## What problem does this address?
The table of content blocks was first released in [Gutenberg 13.3](https://make.wordpress.org/core/2022/05/26/whats-new-in-gutenberg-13-3-0-may-25/) as an experimental block. A few releases later, no major bugs have been reported and we should start considering taking the block out of experimental.

## What is your proposed solution?
Taking the block out of experimental would bring the block closer to shipping in WordPress 6.1.

Apart from removing the experimental flag in `block.json` it would be good to stabilize the dependent selector `__experimentalGetGlobalBlocksByName`, first introduced in #39610.
- @getdave do you see any reason not to stabilize the mentioned selector?
- @ZebulanStanphill @gziolo do you see any blocker not to stabilize the ToC block or other steps that should be done?

## Pending items
- [x] Stabilize the dependent selector `__experimentalGetGlobalBlocksByName`: https://github.com/WordPress/gutenberg/pull/58156
- [ ] https://github.com/WordPress/gutenberg/issues/41173: https://github.com/WordPress/gutenberg/pull/72110
- [ ] https://github.com/WordPress/gutenberg/issues/41031: https://github.com/WordPress/gutenberg/pull/72110
- [x] Add an aria-label attribute to the wrapping nav element: https://github.com/WordPress/gutenberg/pull/71586
- [ ] https://github.com/WordPress/gutenberg/issues/43595: https://github.com/WordPress/gutenberg/pull/72110
- [x] Use the same block toolbar controls as the List block (being able to switch between bullet styles): https://github.com/WordPress/gutenberg/pull/71587

**Comment by ZebulanStanphill** [comment]
My main concern with stabilizing the Table of Contents block is that, if the ideas in #41173 are implemented, it could drastically alter both the attribute schema and saved markup of the block. We might end up creating another backwards compatibility nightmare if we stabilize the block in its current form. I haven't had time to work on prototyping this idea, unfortunately.

The more I think about it, the more I realize that using a Table of Contents outside of the post content is actually the primary use-case for some website designs, so it feels almost wrong to have it be stabilized while it remains unusable there.

There's also the undo history bug described in #41031, which is just rather annoying to deal with. I have no clue how to fix it.

**Comment by bph** [comment]
While you are looking at another iteration on the Table of Content Block, 
I am pulling the issue of missing options to modify how the headlines are displayed from the many comments from the [original PR ](https://github.com/WordPress/gutenberg/pull/29739). 
While testing in PR stage, I felt strongly that there needs to be a way for a user to change the style of the list created. Not everyone appreciates an ordered list, and would rather use bullet points or arrows or something else entirely. 

During the various iterations it [started out as bullet list](https://github.com/WordPress/gutenberg/pull/29739). 
After initial discussions about missing options,  it was actually changed to an ordered list. Although, there were other voices to advocate for more options for the display of the headlines, it did not receive any further consideration,  

- @paaljoachim [commented ](https://github.com/WordPress/gutenberg/pull/29739#issuecomment-1065861880) on it. 
- @juanmaguitar [commented](https://github.com/WordPress/gutenberg/pull/29739#issuecomment-1066492458) on it as well. 

During the discussion, it was not accepted as a needed feature, and argued that it was in the theme developer's domain. 

I still feel that an end user should be able to change the nature of the Table of Content list, via an option in the Sidebar, rather having the need to connect with a developer to modify the theme css. 
In one comment it was mentioned that such settings should be in the [Global Styles and deferred to a future interation.](https://github.com/WordPress/gutenberg/pull/29739#issuecomment-1068514299) 

To be consistent, most core blocks have myriad of options applied to each instance of the block, in addition to the theme.json settings for Global Styles. 

For now, the current Table of Contents block doesn't provide neither block based options nor a theme.json or Global Style implementation to change the display of the Headlines. 

With more options it would shine so much brighter, especially if it makes it into the next major WordPress release.

**Comment by ndiego** [comment]
I 100% agree with @bph and @ZebulanStanphill assessments. This block, while a great start, could use some additional love. To echo some of the previous comments:

1. The largest area of concern to me is the fact that it does not work outside of post content, thereby making the block unusable in the Site Editor. Without a fix here, I anticipate a lot of user confusion and frustration. You are able to add the TOC block to templates, yet it doesn't display anything. 
2. The block needs more style settings to be consistent with other Core blocks. Color, typography, list style, and dimensions would be a good start.
3. There needs to be a way to control the headings that are displayed. A common use case is only to display H2 and H3s, but not the H4, H5, etc.

**Comment by getdave** [comment]
> @getdave do you see any reason not to stabilize the mentioned selector?

As far as I'm aware there is no reason not to stablise this selector. Also noting [it is not widely used](https://github.com/WordPress/gutenberg/search?q=__experimentalGetGlobalBlocksByName).

**Comment by Mamaduka** [comment]
I think styling enhancements aren't necessary blockers for stabilization. These can be done after.

The "Table of Contents outside of the post content" is nice to have but not a blocker, in my opinion. My best guess is that we'll have to use dynamic rendering and parse content blocks to achieve this.

The "undo bug" breaks editor history, so it probably should be fixed before we consider stabilizing the block — recent findings from @kevin940726 on this issue - https://github.com/WordPress/gutenberg/issues/41031#issuecomment-1188544191.

P.S. It would be nice to stress test the block on large posts and see if it affects typing.

**Comment by ndiego** [comment]
Yeah, I am all for stabilizing. I just don't think we should include it in Core until the block is a bit more full-featured. While a great step forward, it feels incomplete.

**Comment by gziolo** [comment]
> Yeah, I am all for stabilizing. I just don't think we should include it in Core until the block is a bit more full-featured. While a great step forward, it feels incomplete.

Technically speaking, the issue is about exposing this block for WordPress Core by removing the experimental flag that ensures Table of Contents block can be used only with the Gutenberg plugin.

> The more I think about it, the more I realize that using a Table of Contents outside of the post content is actually the primary use-case for some website designs, so it feels almost wrong to have it be stabilized while it remains unusable there.

> The "Table of Contents outside of the post content" is nice to have but not a blocker, in my opinion. My best guess is that we'll have to use dynamic rendering and parse content blocks to achieve this.

We need to make the decision whether this block should be general purpose and work everywhere or we limit its usage to the post editor. One way to approach it could by setting in `block.json` the `ancestor` field to `core/post-content`. It probably won't work out of the box, but it shouldn't be too difficult to implement.

In general, it's a challenge to ensure that the Table of Contents block stays in sync when some dynamic blocks change outside of the place where the table of contents gets updated. It's also an issue when using the post editor and you have a Reusable block with heading that can be modified in every place on the site.

**Comment by joedolson** [comment]
The table of contents block also needs to add an `aria-label` attribute to the wrapping `nav` element. There will almost always be other nav elements present on the page (the main navigation, at minimum), so this needs to be made clear. The obvious choice for single post views is just 'Table of Contents', but if it's used outside of a single post view, it'll need further differentiation from other tables of contents.

**Comment by apeatling** [comment]
cc @priethor to see if this is still important for 6.1.

**Comment by ZebulanStanphill** [comment]
I should note that due to being busy with several other things, it is unlikely that I will be able to help much with stabilizing the Table of Contents block, besides reviewing PRs from other people. If anyone wants to tackle the aforementioned issues, feel free to do so, because I simply don't have the time available to do so right now.

If you ask me, the switch back to dynamic rendering and the question of a bullet style option are two things that must be tackled prior to the block being considered stable.

**Comment by priethor** [comment]
I think it is still worth exploring having the ToC block in 6.1. I don't think being limited to the Post Editor is a blocker; to me, it's an opportunity to include it earlier with a limited scope and iterate later on it.

**Comment by ZebulanStanphill** [comment]
My primary concern is not the limitations it would have if merged now, but rather the backwards-compatibility burden it would cause later on, if (or most likely, _when_) the implementation changes. I've already seen several recent cases of core blocks having to make a bunch of compromises due to an inability to update existing static-rendered blocks on the front-end. For example:

- The reverted attempt to remove the wrapping `<div>` from the Button block's markup. #21266, #21923
- Adding a `wp-block-heading` class to the Heading block. #42122
- Difficulties in adding a new default layout mode for the Group block. #42763

**Comment by priethor** [comment]
Bringing this to @michalczaplinski and @ockham's attention as a decision needs to be made for 6.1.

While I'm all in for including this in core feature-wise, apart from that, it hasn't been released in Gutenberg in a stable manner yet and won't be before Beta 1.

**Comment by michalczaplinski** [comment]
Based on all the concerns in the current thread I would not feel comfortable including the Table of Contents block in the 6.1 in its current form.

It seems that we'd at least need to complete the following:
- https://github.com/WordPress/gutenberg/issues/41031 - this is just a bugfix but an annoying one it seems.
- https://github.com/WordPress/gutenberg/issues/41173 - adding support for this likely means switching to dynamic rendering of the block. This means that if we were to stabilize the current (static) version of the block, it would create compatibility problems when adding this fix later on.
- Add the ability to customize the style of the list in the Table of Contents as mentioned by @bph in [this comment above](https://github.com/WordPress/gutenberg/issues/42229#issuecomment-1186211465).

The Feature Freeze for 6.1 is [in exactly 2 weeks](https://make.wordpress.org/core/6-1/#:~:text=for%20the%20release.-,20%20September%202022,-Beta%201.%20From). Are there any volunteers that would like to work on those features so that we could stabilize the ToC block and include it in the 6.1 ? Are there any other issues that would be considered must-haves? 🙂

**Comment by ndiego** [comment]
The TOC block was not included in the 6.1 Beta, so I am going to remove this PR from the 6.1 Project Board.

**Comment by priethor** [comment]
Let's aim to have this for 6.2 🙏

**Comment by NicoHood** [comment]
So is this block still on the roadmap for 6.2? It would be great to have, as its been in development for so long now :)

**Comment by annezazu** [comment]
Pinging 6.2 Editor Tech co-leads @Mamaduka and @ntsekouras as a heads up as this was punted from 6.1. Going to add to 6.2's project board for now for consideration.

**Comment by ntsekouras** [comment]
I haven't worked or followed much of the work for this block, but after reading the described issues, I'd have to echo @michalczaplinski [comment](https://github.com/WordPress/gutenberg/issues/42229#issuecomment-1238754703) 

>Based on all the concerns in the current thread I would not feel comfortable including the Table of Contents block in the 6.1 in its current form.

Since the discussions for `6.1` there was no activity at all for moving this forward, so I don't think we can stabilize it.

What we should do though is try to actively move forward the remaining issues as soon as possible, in order to include it in `6.3`.

**Comment by priethor** [comment]
Let's punt this issue from 6.2, as it's in the same situation as it was before 6.1.

**Comment by mahnunchik** [comment]
Any chance to have ToC block enabled in 6.2?

Question: how to enable ToC block manually? I really need it 😉

**Comment by annezazu** [comment]
It will not be included in 6.2! As you can see above, that decision was made in January. If you want to use it for now, you can use the Gutenberg plugin but please read this article first before considering using the [plugin that's meant for very early adoption](https://wordpress.org/news/2021/04/become-an-early-adopter-with-the-gutenberg-plugin/).

**Comment by mahnunchik** [comment]
@annezazu Thank you. How to understand which version of Gutenberg plugin is bundled with WP by default?

According to this table https://developer.wordpress.org/block-editor/contributors/versions-in-wordpress/ I can install Gutenberg plugin versions from `14.2` to `15.1` for `6.2`

**Comment by joedolson** [comment]
@mahnunchik The version of the editor included in core isn't an exact match for any released version of Gutenberg. Generally speaking, there's a release defined as the final release that will be included in core, but later bug fixes will be merged back into core even while further feature development continues. So version 6.2 contains a version of the block editor that's very similar to Gutenberg version 15.1, but it is not identical.

**Comment by annezazu** [comment]
Looping back on this ahead of 6.3. To focus the discussion and provide an update, the following are likely needed for inclusion in 6.3:

- [Table of Contents: support usage outside of Post Content](https://github.com/WordPress/gutenberg/issues/41173#top)
- [Table of Contents: fix undo history quirk](https://github.com/WordPress/gutenberg/issues/41031#top)
- Add an option to control the headings that are displayed.
- Resolve this autosave issue: https://github.com/WordPress/gutenberg/issues/43595

In the future, more styling options can be added. I'm going to add this to the broader Phase 2 board in case someone can pick the work up.

**Comment by bph** [comment]
Thank you for picking up this thread again, @annezazu 

TBH, the architecture might need refactoring, so a styling of the list is possible, without losing the automatic update with new headlines. At the moment, the TOC block is restricted as a hard-coded ordered list. A conversion to a bulleted list is only possible by using the crutch of converting it to a static list, that then loses the ability to automatically update when the writer adds a headline or change the order of the headline. 

To update the TOC with the updated outline of the post the user is forced either to wait to add the TOC block until the post is fully composed or employ the following ritual: 

- delete the static list block, 
- add the TOC block again that shows the new outline. 
- Covert to static list again 

The user will expect the Table of Content block to update automatically. A restriction to the numeric block is not a suitable trade-off. 

> I think styling enhancements aren't necessary blockers for stabilization. These can be done after. (@mamaduka) 

It is my understanding, - and I have been wrong before many, many times, -  that for more styling options, there would need to be more flexibility in the underlying markup, that won't be backwards compatible, unless the block includes a conversion function, or is a dynamically rendered block. Either way would need to get some consideration before making it into Core.

**Comment by MatzeKitt** [comment]
> At the moment, the TOC block is restricted as a hard-coded ordered list.

Why is this a bad thing? A TOC is, by design, an ordered list, since it’s an ordered structure of headlines in the post. For my understanding, everything else would be semantically incorrect for a TOC.

**Comment by bph** [comment]
Here is the Custom CSS to transform the numbers in an ordered list to bullets: TY @justintadlock 
```
 ol {
  list-style: disc;
}
```
Open the Styles sidebar, scroll all the way down to "Blocks",scroll to 'Table of Content' block, click on 'Additional block CSS', copy/paste the snippet and don't forget to click the "Save" button. Now you can use the Table of Contents block with bullet points without having to convert to a static list. 
<img width="804" alt="Screenshot 2023-05-26 at 17 07 50" src="https://github.com/WordPress/gutenberg/assets/39980/9fec1a85-8a48-45e9-96c8-fbb1f05acede"> (partial table of content in this post [What’s new for developers? (April 2023)](https://developer.wordpress.org/news/2023/04/whats-new-for-developers-april-2023/))

_I just leaving it here, in case there won't be any more styling tools for the Table of Contents block, as a reminder to create a block style for the bullet list and provide a work around for others_

**Comment by richtabor** [comment]
> The block needs more style settings to be consistent with other Core blocks. Color, typography, list style, and dimensions would be a good start.

Just following up that these are now added: 

<img width="281" alt="CleanShot 2023-06-29 at 11 51 42" src="https://github.com/WordPress/gutenberg/assets/1813435/14d932c8-d384-4962-abdb-12479980aae7">

**Comment by richtabor** [comment]
> The table of contents block also needs to add an aria-label attribute to the wrapping nav element. 

> The "Table of Contents outside of the post content" is nice to have but not a blocker, in my opinion. My best guess is that we'll have to use dynamic rendering and parse content blocks to achieve this.

@Mamaduka Is it possible to have it always pull headings from the post content, wherever the block is rendered on the page (within templates or not)? Re: https://github.com/WordPress/gutenberg/issues/41173#issuecomment-1613437159

Seems these are the last two bits to get this in.

**Comment by richtabor** [comment]
> At the moment, the TOC block is restricted as a hard-coded ordered list.

<img width="332" alt="CleanShot 2023-06-29 at 11 56 37" src="https://github.com/WordPress/gutenberg/assets/1813435/e1fd1346-9f06-4cef-9e35-31b4f3cc22d0">

Perhaps we could use the same block toolbar controls as the List block, but apply list style CSS instead of changing the markup.

**Comment by ghost** [comment]
Are the authors Zebullan or Priethor still commited for this block to be stabilized? What course of action is proposed now? Plan A and plan B? There are some open issues. If all cannot be attained at the same time as it seems, what should be prioritized and what should be discarded? Let's discuss this to have a clear path.
Kindly,

(The Details block and the Footnotes block recently landed, and they had their scope limited in the development. So far so good despite the limitations. I just asked myself if that is also the way for the TOC block).

**Comment by annezazu** [comment]
There's a clear path right now. The blocker is someone diving in to do the work right now. For 6.3, we weren't able to get someone in place who could do so.

**Comment by ZebulanStanphill** [comment]
My schedule is very busy lately, so it is unlikely that I will be able to do much here, though I'll try and review any PRs relating to the block if I have the time.

**Comment by Mamaduka** [comment]
@annezazu, @priethor can you update the description and include list of items required for stabilization?

I’ll try to resolve those items before WP 6.4.

**Comment by priethor** [comment]
I've consolidated the remaining items in the issue description. 🚀

**Comment by Mamaduka** [comment]
The #54224 is ready for review. Fixes the undo quirk and page permalink bugs.

**Comment by bph** [comment]
As a[ follow-up on the decision](https://wordpress.slack.com/archives/C055Y7FKS7N/p1694718285665909) that the Table of Content will not be shipping in 6.4, I removed it from the project board

**Comment by phanduynam** [comment]
Looks like the table of contents block is not published on wordpress 6.5.2

**Comment by phanduynam** [comment]
WordPress version 6.5.3 does not have a table of contents block, I recommend adding a Table of Contents block to WordPress 6.6 to quickly create a table of contents based on all the headings on a page. From then on I won't need to install the Table of Contents Plugin anymore.

**Comment by mikachan** [comment]
Would the Table of Contents block be a good candidate for a canonical block plugin? I see it mentioned on this related issue: https://github.com/WordPress/gutenberg/issues/58773. It would still need to be installed as a plugin, but as it would be a block plugin this would be a smoother experience, as they can be installed directly from the inserter.

Edit: I just saw your [comment](https://github.com/WordPress/gutenberg/issues/58773#issuecomment-2126362087) over on the same issue 😄

**Comment by phanduynam** [comment]
> Would the Table of Contents block be a good candidate for a canonical block plugin? I see it mentioned on this related issue: #58773. It would still need to be installed as a plugin, but as it would be a block plugin this would be a smoother experience, as they can be installed directly from the inserter.

There are a total of about 1 million active installs for the Table of Contents plugins, which I think is large enough to include the Table of Contents Block in WordPress core

**Comment by masteradhoc** [comment]
@richtabor any chance to get this stabilized soon?

**Comment by asafm7** [comment]
I don't know if this is a known issue and if it is the place to report it, but when using the Gutenberg TOC on a WooCommerce Single Product template the headings are listed properly, but the items on the TOC aren't linking to them - they are just static `<li>`.

**Comment by annezazu** [comment]
I work with folks on the Woo side and will ping them about this!

**Comment by WunderBart** [comment]
> I don't know if this is a known issue and if it is the place to report it, but when using the Gutenberg TOC on a WooCommerce Single Product template the headings are listed properly, but the items on the TOC aren't linking to them - they are just static `<li>`.

@asafm7, it seems to be behaving like that for me on all templates, with only the Gutenberg plugin enabled (no WooCommerce). For example, if you add it to `Blog Home` it will also render static items. Could you check that on your side, please? 🙏

**Comment by asafm7** [comment]
@WunderBart I tried it in a Single Post template which I considered the most straightforward use case for the TOC block. It doesn't work at all in this use case - nothing is showing on the front end.

In the block description, it says "Headings with HTML anchors will be linked here". I've tried adding HTML Anchors to the headings and it is still not working.

Unless I'm missing something, the editor doesn't seem to automatically add anchors to headings anyway, which makes the usage of the TOC block, even if it did work, a bit tedious.

**Comment by WunderBart** [comment]
> @WunderBart I tried it in a Single Post template which I considered the most straightforward use case for the TOC block. It doesn't work at all in this use case - nothing is showing on the front end.
> 
> In the block description, it says "Headings with HTML anchors will be linked here". I've tried adding HTML Anchors to the headings and it is still not working.
> 
> Unless I'm missing something, the editor doesn't seem to automatically add anchors to headings anyway, which makes the usage of the TOC block, even if it did work, a bit tedious.

Thanks for checking, @asafm7. I'm trying to verify if the TOC block is broken specifically in the WooCommerce context or if it's broken upstream. For me, the buggy behavior I [described above](https://github.com/WordPress/gutenberg/issues/42229#issuecomment-2258319864) is the same regardless of whether I have the WooCommerce plugin active or not. Having said that, I've tested against the `Single Posts` template as you did above, and while it does not render the `Headings with HTML anchors will be linked here` placeholder for me, it still renders static list items instead of links:

| editor | frontend |
| ------------- | ------------- |
| ![localhost_8888_wp-admin_site-editor php_postId=twentytwentyfour%2F%2Fsingle postType=wp_template canvas=edit(Desktop)](https://github.com/user-attachments/assets/b7dcfe8f-cafc-4a85-a68a-edd385ec6170) | ![localhost_8888_howdy-doodie_(Desktop)](https://github.com/user-attachments/assets/661a5580-6b92-4f9d-9899-fca357db5cf3) |

**Comment by asafm7** [comment]
Thanks, @WunderBart. So yes, it seems to be a general template-context issue rather than a WooCommerce issue.

**Comment by TheJeffr0** [comment]
Any updates on the TOC block?

**Comment by michaelh99** [comment]
+1 more for the toc block.

Based on reviews, it looks like using the gutenberg plugin might no longer be a viable workaround

**Comment by dasurus** [comment]
Does the development stop for this block?

**Comment by annezazu** [comment]
A quick update here as I'm reviewing various things in progress for 6.9. This is listed as a block to [consider for inclusion in Core](https://github.com/WordPress/gutenberg/issues/71026) and work has been picked back up in an effort to make that happen with two main PRs: https://github.com/WordPress/gutenberg/pull/71586 & https://github.com/WordPress/gutenberg/pull/71587. As with the other blocks under consideration, this will be placed behind an experiment and evaluated for inclusion in 6.9 as we get closer.

**Comment by t-hamano** [comment]
Let's add this issue to the 6.9 project board so we don't forget about it.

**Comment by t-hamano** [comment]
[A PR was submitted to try dynamic rendering](https://github.com/WordPress/gutenberg/pull/72110) to stabilize the ToC block. However, there are issues and concerns that need to be resolved, so it couldn't be released in the 6.9 cycle. Let's aim for the 7.0 release.

**Comment by t-hamano** [comment]
[The 7.1 roadmap proposes stabilizing this block](https://make.wordpress.org/core/2026/06/19/roadmap-to-7-1/#:~:text=Table%20of%20Contents%20block%2C), but this task is not active. It is likely difficult to ship in 7.1 as there are still many technical challenges to address. My understanding is that to move this forward, we need technical feedback and discussion on the following points: 

- Static rendering vs. dynamic rendering?
- What and how should be displayed in the template?
- How should automatic generation of heading anchors be handled?
- How to extract heading elements other than heading blocks?

**Comment by mtoensing** [comment]
Just a humble suggestion: maybe we could use some of the principles from https://github.com/mtoensing/SimpleTOC ? I am one of the authors, but it appears that it is close to the idea of this block here with the same challenges?

**Comment by annezazu** [comment]
@scruffian any chance your crew will be able to get to this? If not, I agree with punting. It feels tight as it is. 

@mtoensing if you’d like to contribute here, please do! It would be awesome to learn from your approach and have some code contributions. Thanks for chiming in to say something.

**Comment by Mamaduka** [comment]
> Static rendering vs. dynamic rendering?

I think, based on the latest experiments, dynamic rendering made more sense. The static doesn't work when a block is placed inside templates and template parts.

IIRC, there were some reservations about reparsing the entire content inside the block, but this might be easier/faster with the new block parser.

cc @dmsnell, @sirreal, we could use any pointers here?

**Comment by mtoensing** [comment]
@annezazu 
SimpleTOC’s approach has been to keep the block narrow: generate a minimal nested list from post headings, avoid JS/CSS by default, rely on native block/theme styling where possible, and expose only a few user-facing controls such as heading depth, ordered/bullet list, and optional wrapper/ARIA settings.

From my experience, the most important decisions for core are:

1. Prefer dynamic rendering for template/template-part support. Nothing more because then it gets complicated for less than 20 % of the use cases.
2. Keep saved markup minimal to reduce future compatibility burden.
3. Treat heading extraction as a server-side responsibility where possible.
4. Avoid adding custom styling systems; lean on existing block supports/group styling.
5. Make heading depth and list type core options, because users expect those without custom CSS.

I’d be happy to help test or contribute parts of this, especially around rendering, heading extraction, depth control, and compatibility lessons from SimpleTOC.

**Comment by druxstr** [comment]
If the block is to be stabilized, I suggest we add #52378 to the pending items, as it has some accessibility considerations.

**Comment by t-hamano** [comment]
> Treat heading extraction as a server-side responsibility where possible.

Does this mean that the block is server-side rendered in the editor? If so, even if we add a heading to the post editor, that heading may not be added to the ToC block unless we save the post. This is because server-side rendering means referring to the post content in the database.

If we can accept this point, I think it is worth trying, as we can consolidate all the logic on the server-side and enjoy the benefits of the HTML processor.

**Comment by adrianduffell** [comment]
> Does this mean that the block is server-side rendered in the editor? If so, even if we add a heading to the post editor, that heading may not be added to the ToC block unless we save the post. This is because server-side rendering means referring to the post content in the database.

Server-side rendering seems like the best foundation for this block, but with this trade-off the editor isn't in sync until the post is saved. A few thoughts on ways to help alleviate this:

* Could it be synced with autosave content? I believe this is updated every minute, so could increase the frequency of syncs.
* Adding a notice that the ToC may show outdated content until the next save.
* Displaying a static placeholder in the editor. Not a fan of this overall, but may be worth it as a stopgap to have this ready for 7.1.

**Comment by sirreal** [comment]
> cc [@dmsnell](https://github.com/dmsnell), [@sirreal](https://github.com/sirreal), we could use any pointers here?

I've looked over https://github.com/WordPress/gutenberg/pull/72110 as a dynamic rendering implementation. It does look like a place where [the WP_Block_Processor](https://developer.wordpress.org/reference/classes/wp_block_processor/) is suitable. Migrating that implementation will likely be more efficient for extracting the table of contents information.

**Comment by getdave** [comment]
Catching up here...

> Server-side rendering seems like the best foundation for this block, but with this trade-off the editor isn't in sync until the post is saved. A few thoughts on ways to help alleviate this:

We have a precedent in the Nav block for having a PHP class that acts as a central source of truth for an algorithm that needs to remain consistent across _both_ editor and front of site rendering. We expose this over a block editor specific REST endpoint for the Editor whilst the front end code calls it directly.

The result is that we have a single source of truth for behaviour and that avoids drift over time.

We should be able to follow that path here as well, with the proviso that we might need to pass some data from the Editor as context so that the algorithm can operate dynamically and function correctly even if the post isn't saved to the database yet.

This would introduce some lag but I think for a block like this that is perfectly acceptable. It's not critical that the preview updates instantly. It's more important that it's accurate and consistent.

**Comment by getdave** [comment]
As there is a lot of required prior context on this block I've tried to collate this and then generate **a set of User Stories** that document the expected behaviour from a user point of view.

I've added these to a collapsed section in the Issue description above. It would be valuable if we could validate these so that they can act as a baseline for all contributors (including Agents).

**Comment by getdave** [comment]
> Catching up here...
> 
> > Server-side rendering seems like the best foundation for this block, but with this trade-off the editor isn't in sync until the post is saved. A few thoughts on ways to help alleviate this:
> 
> We have a precedent in the Nav block for having a PHP class that acts as a central source of truth for an algorithm that needs to remain consistent across _both_ editor and front of site rendering. We expose this over a block editor specific REST endpoint for the Editor whilst the front end code calls it directly.
> 
> The result is that we have a single source of truth for behaviour and that avoids drift over time.
> 
> We should be able to follow that path here as well, with the proviso that we might need to pass some data from the Editor as context so that the algorithm can operate dynamically and function correctly even if the post isn't saved to the database yet.
> 
> This would introduce some lag but I think for a block like this that is perfectly acceptable. It's not critical that the preview updates instantly. It's more important that it's accurate and consistent.

> I've looked over https://github.com/WordPress/gutenberg/pull/72110 as a dynamic rendering implementation. It does look like a place where [the WP_Block_Processor](https://developer.wordpress.org/reference/classes/wp_block_processor/) is suitable. Migrating that implementation will likely be more efficient for extracting the table of contents information.

To confirm you are proposing using the `WP_Block_Processor` to parse the post data's headings and anchors?

**Comment by Mamaduka** [comment]
> We expose this over a block editor specific REST endpoint for the Editor whilst the front end code calls it directly.

Probably a bit much for the ToC block. I won't worry too much about the block editor rendering at the moment; it's working as an okay preview.

I think the largest initial work would be to use the `WP_Block_Processor` to parse the post data's heading blocks and their anchors.

**Comment by t-hamano** [comment]
A point to note is that a block may contain multiple heading "elements", or heading elements may be contained within custom HTML blocks or classic blocks. While extracting only heading blocks might be sufficient for now, in the future, it may be necessary to use `WP_HTML_Tag_Processor` in addition to `WP_Block_Processor`.

**Comment by sirreal** [comment]
I'm _not_ deeply familiar with the efforts so far, I'm trying to advise from the perspective of the available parsing/processing tools.

> To confirm you are proposing using the WP_Block_Processor to parse the post data's headings and anchors?

If we take the dynamic rendering approach and content needs to be parsed to generate the table of contents, absolutely. The block processor is the first approach I would try.

> it may be necessary to use WP_HTML_Tag_Processor in addition to WP_Block_Processor.

The tag processor may be sufficient on its own if we're looking at typical serialized post content like this:

```html
<!-- wp:heading {"level":1} -->
<h1 class="wp-block-heading">ok</h1>
<!-- /wp:heading -->

<!-- wp:html -->
<h2>yep
<!-- /wp:html -->
```

**Comment by getdave** [comment]
> Probably a bit much for the ToC block.

Can you explain why you feel that? There might be an argument that for purposes of expediency we could go with separate editor and front end code paths for initial release and look to normalize later. 

That said, in my experience in this repo (and with blocks in particular) maintaining x2 versions of the same algorithm in separate places is a recipe for drift and errors.

-----

**Quick ask from me**: please could we confirm we are aligned on the _expectations_ of this block from [a user experience perspective as per this comment](https://github.com/WordPress/gutenberg/issues/42229#issuecomment-4902768019). It's easy for context to get lots in giant threads and it will be easier to implement the right solution if we're all aligned. Thanks in advance 🙏

**Comment by Mamaduka** [comment]
> Can you explain why you feel that?

Similar to other blocks that derive values from content, if content or blocks aren't in the store, the editor can display a placeholder. This is more common when a block is placed inside a template, template part or pattern. I think a simple placeholder is more appropriate for ToC.

tl;dr: There's no need to focus on editor display right now. Getting a server-side renderer with `WP_Block_Processor` is more important.

**Comment by t-hamano** [comment]
With 7.1 Beta1 just around the corner and no reviewable PRs ready yet, let's punt this to 7.2.

**Comment by getdave** [comment]
@t-hamano I was just coming here to post something similar.

My read on the existing situation is that we're still waiting on a well tested solution for server side rendering the block that also accounts for the situation where the block appears in a template (example: you want to add to the `Posts` template so that the ToC appears on _every_ Post to avoid having to manually add it to every Post in the database). 

Given the cut off I don't feel confident that we have time to implement a well tested solution.

Moreover, the block is lacking automated test coverage generally which means it's hard to 

- iterate with confidence
- feel confident in its overall stability for a major release.

In terms of dynamic rendering, I have pushed up https://github.com/WordPress/gutenberg/pull/80098 but I would be far from confident in it merging prior to the 7.1 cut off. 



## Proposal for 7.2 cycle

Whilst not stable enough for 7.1, I feel the block is not too far from being shippable. With the right planning it should be achievable for 7.2.

Here's what I'm proposing:

- [Add test coverage for the behaviour we want the final block to have](https://github.com/WordPress/gutenberg/pull/80199) in the 7.2 cycle - this documents the desired behaviour _in code_ and avoids contributors needing to build/rebuild a mental model from disperate Github threads/comments. For purposing of merging code some tests can be stubs with "todo" attached to document expected behaviour. 
- Merge a solution which handles the basics of dynamic rendering on the PHP side. Edge cases can follow up.
- Investigation spike into feasibility of Editor and Front end rendering sharing the same code paths for the headings calculation algorithm.
- Algorithm implementation - either:
    - implement co-located algorithm solution, or...
    - add tests to provide strong confidence that separate Editor and PHP-based resolution code paths are identical
- Fix edge cases such as 
   - paginated posts
   - query loop 
   - placement in unsuitable templates (e.g. `archive-` type templates).
   - ...etc
- Final testing and UX polish including adding controls for heading level...etc.

How does this sound and is there anything anyone would like to add?

I want to state my thanks to contributors for all the effort that's gone into this block so far. Let's try and get this over the line for 7.2.

**Comment by t-hamano** [comment]
@getdave Thank you for summarizing it!

In my personal opinion, the most crucial point is to robustly handle the extraction of expected headings on the server-side. Specific UIs or options are not necessary for the MVP.

**Comment by getdave** [comment]
## Update

This effort is now targeting WordPress 7.2. 

Style settings, heading level filtering, dynamic server-side rendering (#80318), and dropping the block's saved static markup (#80404) have all merged.

The remaining work is to determine where the _Editor's_ heading level calculation algorithm should run and to weight the trade offs of the various approaches we've uncovered - namely:

- run server side and render via editor's existing server-side render hooks
- run on Editor (client) side
- run server side using a custom REST endpoint and have the editor send its current, unsaved content to get the headings

Each has pros and cons and it's important we fully understand these before making a decision.

Template context, referenced-entity headings, third-party heading sources, and the filterable resolver (#80787, #80788) all depend on it and haven't started.

### Where we're blocked

Discussion on the REST endpoint sub-issue (#80786) raised a question about whether a dedicated endpoint is the right call, versus reusing the editor's existing server-side render mechanism. 

That produced a feasibility spike (#80794) and a working PR (#80796) that reuses the existing server-side rendering hooks for the editor preview. 

Whilst appealing in its simplicity, unfortunately that spike surfaced a UX problem I don't think we can accept. Because it reads from the post as saved in the database, **the block's preview only updates on save, not as the author edits**. This manifests as the following:

- A new,  unsaved post shows no preview until first save. 
- A saved draft's preview stays stale after an edit until autosave or a manual save catches up,
- A published post's preview can stay stale until the author clicks `Update`.

This doesn't feel like an acceptable UX to me. 

Another option is maintaining a duplicate heading level algorithm on the _Editor_ side. Whilst the overhead might be low it still raises the possibility of implementation drift over time. More importantly however, all extensibility would then need to be implemented _twice_ (in both JS and PHP) which would then also require 3rd party authors to register extensions _twice_. This is clearly suboptimal.

The final option, is having the editor send the post content to the server via dedicated REST endpoint and [reusing the same heading algorithm code](https://github.com/WordPress/gutenberg/pull/80372) as used by the PHP front of site rendering to provide a response for the Editor to render. 

This solves the stale rendering and extensibility problems outlined above, but comes with a potential latency cost resulting from having to send post content over the network on every debounced key stroke.

### My recommendation

I am still inclined towards the last option of a custom REST endpoint as it provides the best _user_ (and 3rd party developer) experience even if it is harder for us as Core developers to implement.

I fully recognise the latency issue but we may be able to offset this by 

- sending only a filtered version of the contents (only registered headings?)
- being aggressive with our debouncing (_instant_ previews for the ToC block aren't necessary)

In order for this block to move forward we need to come to a decision on which route to take. I'd very much appreciate others reviewing the above and offering their recommendation - @t-hamano, @Mamaduka, @tyxla, @scruffian.

**Comment by Mamaduka** [comment]
Thanks, @getdave!

> Whilst appealing in its simplicity, unfortunately that spike surfaced a UX problem I don't think we can accept. Because it reads from the post as saved in the database, the block's preview only updates on save, not as the author edits. This manifests as the following

It's not clear to me how a dedicated REST endpoint can fix this case. It should read the post from somewhere to parse the headings, and usually that's the post from the DB, which will have the same issue.

Yes, we can send edited content over the wire, but we can do the same with the existing SSR. Both have to use `POST` requests to account for the large payload, which is often a problem with `GET`.

**Comment by adrianduffell** [comment]
Excited to see this progressing. Thanks, @getdave !

With all the options seeming to have tradeoffs, I like the spike’s approach best for its simplicity. I think polling the REST API on every debounced keystroke could be a sizeable network/performance tax to pay, especially for large posts. I imagine there's further unknown complexities around this.

It might be possible to alleviate the UX concerns somewhat with a short status message for the block, to at least give users better framing around it. Eg:

State | Status Message
-- | --
A new, unsaved post shows no preview until first save | Table of Contents will be generated when the post is saved as a draft, or published
A saved draft’s preview stays stale after an edit until autosave or a manual save catches up | Table of Contents last generated 30 seconds ago
A published post’s preview can stay stale until the author clicks Update | Table of Contents will be re-generated when the post is updated

**Comment by tyxla** [comment]
I agree with @Mamaduka here. I think this is a problem with the spike, not with server-side rendering. Unless I'm missing something, the preview is stale because we pass `post_id` and read the saved post content. The `block-renderer` route already accepts POST, and `ServerSideRender` has an `httpMethod` prop, so we can send the unsaved content there too. I'd say both server-side options cost ~ the same in latency.

So the question for me is how much new public API we want to add. I'd very much prefer to reuse what we have. The one thing to keep in mind is that the endpoint validates `attributes` with `additionalProperties: false`, so we need a plan for sending the content - perhaps a good topic for #80786 before we build a new endpoint.

Regarding extensibility: if #80787 stays declarative (based on block name), we could expose it to the editor, and third parties would register only once. Only #80788 really needs two implementations. So the end result may be simpler than it looks at first glance.

**Comment by getdave** [comment]
> It should read the post from somewhere to parse the headings, and usually that's the post from the DB, which will have the same issue.

It would need to come from the Editor post content otherwise (as you say) it would just have the same issue as the SSR approach. 

> Excited to see this progressing. Thanks, @getdave !

You are welcome 👍 Thanks for jumping in here too.

> It might be possible to alleviate the UX concerns somewhat with a short status message for the block

My personal opinion (lightly held) is that this would be sub par UX. The ideal user experience should be a block that just works and doesn't require reading status messages to understand why it's not working. 

> ServerSideRender has an httpMethod prop, so we can send the unsaved content there too

I wasn't aware of that. I'd like to see the spike PR updated with that approach to see if it resolves the issue with stale content. I'm not entirely clear _how_ it might resolve it yet, but I trust your instinct here and therefore it's worth exploring.

> So the question for me is how much new public API we want to add

💯 agreed on this. I don't want to add new APIs _if_ SSR work in a way which can provide an optimal user experience. 

> Regarding extensibility: if https://github.com/WordPress/gutenberg/issues/80787 stays declarative (based on block name), we could expose it to the editor,

Yes that should work unless I"m missing another extensibility requirement? AFAIK there should be no huge need to replace the algorithm wholesale so perhaps this will prove sufficient.


-----

Recommended next step: modify the current [spike](https://github.com/WordPress/gutenberg/issues/80794) PR to test [@tyxla's idea](https://github.com/WordPress/gutenberg/issues/42229#issuecomment-5602487465) on SSR.
